third_party/move/move-compiler-v2/src/pipeline/ability_checker.rs

```diff
@@ -0,0 +1,256 @@
+// on the type parameter
+fn check_struct_inst(
+    target: &FunctionTarget,
+    mid: ModuleId,
+    sid: StructId,
+    insts: &[Type],
+    loc: &Loc,
+) -> AbilitySet {
+    let ty_params = target.get_type_parameters();
+    ty::check_struct_inst(
+        mid,
+        sid,
+        insts,
+        Some(loc),
+        ty_param_abilities(&ty_params),
+        get_abilities(target),
+        |loc, msg| target.global_env().error(loc, msg),
+        true,
+    )
+}
+
+fn check_fun_inst(target: &FunctionTarget, mid: ModuleId, fid: FunId, inst: &[Type], loc: &Loc) {
+    let qid = QualifiedId {
+        module_id: mid,
+        id: fid,
+    };
+    let fun_env = target.global_env().get_function(qid);
+    for (param, ty) in fun_env.get_type_parameters().iter().zip(inst.iter()) {
+        let required_abilities = param.1.abilities;
+        let given_abilities = check_instantiation(target, ty, loc);
+        // todo: which field, why
+        if !required_abilities.is_subset(given_abilities) {
+            target.global_env().error(loc, "invalid instantiation")
+        }
+    }
+}
+
+// checks whether the type is instantiated properly, i.e.,
+// - the type arguments given to the struct are instantiated properly
+// - the type arguments satisfy the ability constraints defined on the struct generics
+// and returns the abilities of the given type
+// `ty_params` specify the abilities held by type params
+pub fn check_instantiation(target: &FunctionTarget, ty: &Type, loc: &Loc) -> AbilitySet {
+    let ty_params = target.get_type_parameters();
+    ty::check_instantiation(
+        ty,
+        Some(loc),
+        ty_param_abilities(&ty_params),
+        get_abilities(target),
+        |loc, msg| target.global_env().error(loc, msg),
+        true,
+    )
+}
+
+pub struct AbilityChecker();
+
+impl FunctionTargetProcessor for AbilityChecker {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let target = FunctionTarget::new(fun_env, &data);
+        for bytecode in target.get_bytecode() {
+            check_bytecode(&target, bytecode)
+        }
+        data
+    }
+
+    fn name(&self) -> String {
+        "AbilityChecker".to_owned()
+    }
+}
+
+fn check_bytecode(target: &FunctionTarget, bytecode: &Bytecode) {
+    let loc = target.get_bytecode_loc(bytecode.get_attr_id());
+    match bytecode {
+        // drop of dst during the assignment has been made explicit
+        // so we don't check it here, plus this could be an initialization
+        Bytecode::Assign(_, src, _, kind) => {
+            if matches!(kind, AssignKind::Copy | AssignKind::Store) {
+                check_copy_for_temp_with_msg(target, *src, &loc, "cannot copy");
```

Comment:
`AssignKind::Inferred` should be unreachable,  since this pass is after the live var transformation where inferred assignments have been converted to copy or move. I will make it panics if this happens.